[WebIDL] Ensure [Reflect] extended attribute value is provided as a string when used with [Enumerated]

rahimabdi opened the pull request:
#### fe2e7fecaef0f0be223945f38b2323ccdba013df
<pre>
[WebIDL] Ensure [Reflect] extended attribute value is provided as a string when used with [Enumerated]
<a href="https://bugs.webkit.org/show_bug.cgi?id=296861">https://bugs.webkit.org/show_bug.cgi?id=296861</a>

Reviewed by Darin Adler.

When [Reflect] is specified with [Enumerated] (e.g. [Reflect=&quot;someAttr&quot;]),
ensure the value is treated as a proper string. For hyphenated attribute values,
the value is normalized for getter/setter codegen (e.g. &quot;aria-orientation&quot;
becomes aria_orientationAttr).

This PR also updates CodeGenerator::UnquoteStringLiteral() to ensure arguments are
passed consistently.

* Source/WebCore/bindings/scripts/CodeGenerator.pm
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp
* Source/WebCore/bindings/scripts/test/TestObj.idl

Canonical link: <a href="https://commits.webkit.org/298277@main">https://commits.webkit.org/298277@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bed45236949faa4e49071cec06f8877abb24929

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114874 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34614 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25083 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121020 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65575 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35246 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87302 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117822 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28085 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103148 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67694 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27262 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21269 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64680 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97467 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21384 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124216 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41872 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31288 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96105 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42243 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99338 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95890 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24417 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41082 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18919 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37925 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41749 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47271 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41303 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44618 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43046 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5680,7 +5680,12 @@ sub GenerateAttributeGetterBodyDefinition
         if ($attribute->extendedAttributes->{Reflect} && $attribute->extendedAttributes->{Enumerated}) {
             my $attributeName = $attribute->extendedAttributes->{Reflect};
-            $attributeName = $attribute->name if !$attributeName || $attributeName eq "VALUE_IS_MISSING";
+            if (!$attributeName || $attributeName eq "VALUE_IS_MISSING") {
```
Potentially should use UnquoteStringLiteral from CodeGenerator.pm

rahimabdi replied:
Since `CodeGenerator::UnquoteStringLiteral()` checks if the first character is a quote, and then removes them, it may be more work inside this block because the attributeName could come from IDL declaration without quotes rather than value of [Reflect]; for example, `[Reflect] DOMString attr` uses attr. I think I'd have to check if the `attributeName` I'm passing in started with a quote which would complicate this logic but please correct me if I'm misunderstanding (as well as checking against undefined and VALUE_IS_MISSING).

(I did try using it because got a "Identifier 'CodeGenerator=HASH(0x11e1cf910)' should be a string literal" error because the argument didn't start with a quote. I suppose I could check if a value is provided for [Reflect] and if not, check the attribute name in the IDL declaration, etc. but it may be too complicated here.)

rahimabdi replied:
I think it's working now (spent a little too much time trying to figure out why it was passing the codegenerator object rather than reflect string!). Thanks @lukewarlow.

lukewarlow commented on Source/WebCore/html/HTMLIFrameElement.idl:
```diff
@@ -33,7 +33,7 @@
     [Reflect, CEReactions=NotNeeded] attribute DOMString height;
     [CEReactions=NotNeeded, ImplementedAs=referrerPolicyForBindings, ReflectSetter] attribute [AtomString] DOMString referrerPolicy;
-    [CEReactions=NotNeeded, EnabledBySetting=LazyIframeLoadingEnabled, Reflect, Enumerated=HTMLIFrameElementLoadingValues] attribute DOMString loading;
+    [CEReactions=NotNeeded, EnabledBySetting=LazyIframeLoadingEnabled, Reflect="loading", Enumerated=HTMLIFrameElementLoadingValues] attribute DOMString loading;
```
This isn't needed because loading is the default.

rahimabdi replied:
Got it, thanks.

lukewarlow commented on Source/WebCore/html/HTMLImageElement.idl:
```diff
@@ -43,7 +43,7 @@
     [CEReactions=NotNeeded, ImplementedAs=referrerPolicyForBindings, ReflectSetter] attribute [AtomString] DOMString referrerPolicy;
     [CEReactions=NotNeeded, ReflectSetter] attribute [AtomString] DOMString decoding;
-    [CEReactions=NotNeeded, EnabledBySetting=LazyImageLoadingEnabled, Reflect, Enumerated=HTMLImageElementLoadingValues] attribute DOMString loading;
+    [CEReactions=NotNeeded, EnabledBySetting=LazyImageLoadingEnabled, Reflect="loading", Enumerated=HTMLImageElementLoadingValues] attribute DOMString loading;
```
ditto

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5679,8 +5679,16 @@ sub GenerateAttributeGetterBodyDefinition
+            if (!defined $reflect || $reflect eq "VALUE_IS_MISSING") {
+                $attributeName = $attribute->name;
+            } 
+            else {
```
```suggestion
            } else {
```

darinadler reviewed (approved):
No change to any actual IDL file? This doesn't affect any actual DOM?

rahimabdi committed:
[WebIDL] Ensure [Reflect] extended attribute value is provided as a string when used with [Enumerated]
https://bugs.webkit.org/show_bug.cgi?id=296861

Reviewed by Darin Adler.

When [Reflect] is specified with [Enumerated] (e.g. [Reflect="someAttr"]),
ensure the value is treated as a proper string. For hyphenated attribute values,
the value is normalized for getter/setter codegen (e.g. "aria-orientation"
becomes aria_orientationAttr).

This PR also updates CodeGenerator::UnquoteStringLiteral() to ensure arguments are
passed consistently.

* Source/WebCore/bindings/scripts/CodeGenerator.pm
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp
* Source/WebCore/bindings/scripts/test/TestObj.idl

Canonical link: https://commits.webkit.org/298277@main
